core/src/main/java/org/elasticsearch/transport/TcpTransport.java

```diff
@@ -580,17 +585,27 @@ private void closeAndNotify(DiscoveryNode node, NodeChannels nodeChannels, Strin
     /**
      * Disconnects from a node if a channel is found as part of that nodes channels.
      */
-    protected final void disconnectFromNodeChannel(final Channel channel, final Exception failure) {
+    protected final void disconnectFromNodeChannel(final Channel channel, final String reason) {
         threadPool.generic().execute(() -> {
             try {
                 try {
                     closeChannels(Collections.singletonList(channel));
                 } finally {
-                    for (DiscoveryNode node : connectedNodes.keySet()) {
-                        if (disconnectFromNode(node, channel, ExceptionsHelper.detailedMessage(failure))) {
+                    for (Map.Entry<DiscoveryNode, NodeChannels> entry : connectedNodes.entrySet()) {
+                        if (disconnectFromNode(entry.getKey(), channel, reason)) {
                             // if we managed to find this channel and disconnect from it, then break, no need to check on
                             // the rest of the nodes
-                            break;
+                            // Removing it here from open connections is paranoia since the close handler should deal with this.
+                            openConnections.remove(entry.getValue());
+                            return;
```

Comment:
Are you sure you want this? Won't this cause any exception thrown by `closeChannels` to be lost?